### ISSUE TITLE
How to modify element in an OFX object and write them back out?

### ISSUE THREAD

**Issue by pallavg** [csingley/ofxtools#180]
I need to add tickers to certain 401k funds in the OFX obtained from Fidelity to get beancount-import to work properly for my beancount ledger.

I try the following:
1. Parse the OFX file.
2. Convert it to OFX data model.
3. Iterate over the securities and for those that have ticker as None, set it to a string value.
4. Convert the OFX data model back to XML tree.
5. Create header and the two to a file.

Unfortunately, in the output, the ticker is not modified/present. I can see the object is modified but looks like these changes are not being propagated to the overall model. Can someone please show me a small example of how to modify the data members? The documentation only talks about creating OFX data from scratch. Thanks.

Here is the code snippet:

```
 ofx_file = Path(__file__).parent.parent.joinpath('data', 'q1.ofx')

    parser = OFXTree()
    with open(ofx_file, 'rb') as f_in:
        parser.parse(f_in)

    ofx = parser.convert()

    # Add tickers for missing securities based on Fidelity's website.
    for sec in ofx.securities:
        if sec.ticker is None:
            sec.ticker = 'NON40' + sec.secinfo.secid.uniqueid
            print('Set ticker for', sec.secinfo.secid.uniqueid, 'to', sec.ticker)

    for sec in ofx.securities:
        print(sec.ticker)
        print(sec)

    root = ofx.to_etree()

    header = str(make_header(version=220))
    message = ET.tostring(root).decode()
    f = ofx_file.with_name(ofx_file.stem + '_new.ofx')
    with open(f, 'wt') as f_out:
        f_out.write(header + message)
```
```

Output:

Set ticker for O32T to NON40O32T
Set ticker for OT4M to NON40OT4M
Set ticker for O32R to NON40O32R
Set ticker for OXPW to NON40OXPW
Set ticker for OXPU to NON40OXPU
NON40O32T
<MFINFO(secinfo=<SECINFO(secid=<SECID(uniqueid='O32T', uniqueidtype='CUSIP')>, secname='BTC S&P 500 INDEX', fiid='O32T', unitprice=Decimal('11.5435290'), dtasof=datetime.datetime(2023, 12, 19, 21, 0, tzinfo=<UTC>), memo='Market close as of 12/19/2023;BTC S&P 500 INDEX')>, mftype='OTHER')>
NON40OT4M
<MFINFO(secinfo=<SECINFO(secid=<SECID(uniqueid='OT4M', uniqueidtype='CUSIP')>, secname='BTC US DEBT', fiid='OT4M', unitprice=Decimal('12.1787420'), dtasof=datetime.datetime(2023, 12, 19, 21, 0, tzinfo=<UTC>), memo='Market close as of 12/19/2023;BTC US DEBT')>, mftype='OTHER')>
NON40O32R
<MFINFO(secinfo=<SECINFO(secid=<SECID(uniqueid='O32R', uniqueidtype='CUSIP')>, secname='BTC S&P 500 INDEX', fiid='O32R', unitprice=Decimal('11.5412790'), dtasof=datetime.datetime(2023, 12, 19, 21, 0, tzinfo=<UTC>), memo='Market close as of 12/19/2023;BTC S&P 500 INDEX')>, mftype='OTHER')>
NON40OXPW
<MFINFO(secinfo=<SECINFO(secid=<SECID(uniqueid='OXPW', uniqueidtype='CUSIP')>, secname='BTC ACWI EX US IMI', fiid='OXPW', unitprice=Decimal('17.7495280'), dtasof=datetime.datetime(2023, 12, 19, 21, 0, tzinfo=<UTC>), memo='Market close as of 12/19/2023;BTC ACWI EX US IMI')>, mftype='OTHER')>
NON40OXPU
<MFINFO(secinfo=<SECINFO(secid=<SECID(uniqueid='OXPU', uniqueidtype='CUSIP')>, secname='BTC US DEBT', fiid='OXPU', unitprice=Decimal('12.1818410'), dtasof=datetime.datetime(2023, 12, 19, 21, 0, tzinfo=<UTC>), memo='Market close as of 12/19/2023;BTC US DEBT')>, mftype='OTHER')>
FSMAX
<MFINFO(secinfo=<SECINFO(secid=<SECID(uniqueid='315911743', uniqueidtype='CUSIP')>, secname='FIDELITY EXTENDED MARKET INDEX FUND', ticker='FSMAX', unitprice=Decimal('77.8600000'), dtasof=datetime.datetime(2023, 12, 20, 8, 30, 11, tzinfo=<UTC>), currency=<CURRENCY(currate=Decimal('1.000'), cursym='USD')>)>, mftype='OTHER', dtyieldasof=datetime.datetime(2023, 12, 20, 8, 30, 11, tzinfo=<UTC>))>
FDRXX
<MFINFO(secinfo=<SECINFO(secid=<SECID(uniqueid='316067107', uniqueidtype='CUSIP')>, secname='FIDELITY GOVERNMENT CASH RESERVES', ticker='FDRXX', dtasof=datetime.datetime(2023, 12, 20, 8, 30, 11, tzinfo=<UTC>), currency=<CURRENCY(currate=Decimal('1.000'), cursym='USD')>)>, mftype='OTHER', dtyieldasof=datetime.datetime(2023, 12, 20, 8, 30, 11, tzinfo=<UTC>))>
VNQI
<STOCKINFO(secinfo=<SECINFO(secid=<SECID(uniqueid='922042676', uniqueidtype='CUSIP')>, secname='VANGUARD GLOBAL EX-US REAL ESTATE INDEX FD ETF', ticker='VNQI', dtasof=datetime.datetime(2023, 12, 20, 8, 30, 11, tzinfo=<UTC>), currency=<CURRENCY(currate=Decimal('1.000'), cursym='USD')>)>, stocktype='COMMON', dtyieldasof=datetime.datetime(2023, 12, 20, 8, 30, 11, tzinfo=<UTC>))>
VTI
<STOCKINFO(secinfo=<SECINFO(secid=<SECID(uniqueid='922908769', uniqueidtype='CUSIP')>, secname='VANGUARD INDEX FDS VANGUARD TOTAL STK MKT ETF', ticker='VTI', dtasof=datetime.datetime(2023, 12, 20, 8, 30, 11, tzinfo=<UTC>), currency=<CURRENCY(currate=Decimal('1.000'), cursym='USD')>)>, stocktype='COMMON', dtyieldasof=datetime.datetime(2023, 12, 20, 8, 30, 11, tzinfo=<UTC>))>
VCIT
<STOCKINFO(secinfo=<SECINFO(secid=<SECID(uniqueid='92206C870', uniqueidtype='CUSIP')>, secname='VANGUARD INTERMEDIATE TERM CORPORATE BOND ETF', ticker='VCIT', dtasof=datetime.datetime(2023, 12, 20, 8, 30, 11, tzinfo=<UTC>), currency=<CURRENCY(currate=Decimal('1.000'), cursym='USD')>)>, stocktype='COMMON', dtyieldasof=datetime.datetime(2023, 12, 20, 8, 30, 11, tzinfo=<UTC>))>
```

**Comment by csingley** [comment]
TL;DR you almost got it - you actually meant to say this:
```
sec.secinfo.ticker = 'NON40' + sec.secinfo.secid.uniqueid
```
------

First let me just say how much it pleases me to see you really leaning into the isomorphism between SGML and the Python object model presented by `ofxtools`.  But I kinda have to wonder - if it's really SGML that you want, why not just stick with the intermediate representation in XML?  ElementTree has XPath support.

I suspect the answer here is that you find ergonomic the quality-of-life shortcuts I added that provide proxy access to attributes of objects nested deeper in the tree... and those shortcuts, my friend, are _not_ bijective... they are read only.

https://github.com/csingley/ofxtools/blob/master/ofxtools/models/base.py#L540-L549

So, as I sometimes try to point out within the docs, the isomorphism means that often your best avenue for nailing down inquiries lies in reference to the OFX spec... consultation of which will reveal that, in fact, `securities` are not really a thing in OFX; that's another shortcut of my own fabrication... as can be seen by you getting back `MFINFO`, `STOCKINFO` etc. when you iterate over what is really a pointer to a `SECLIST` container.

You might be looking at this through OOP-colored glasses and thinking "well surely `MFINFO` is a subclass of `SECINFO`" ... kind of, kind of... but as you can see from the `repr()`, the `MFINFO` is the container, and the `SECINFO` is the contained element nested within... so the "MRO", if you will, proceeds from outer scope to inner, rather than the other way around as it does in Python... which, parenthetically, is the reason that my little shortcut hack works so neatly to offer proxy access to attributes of subelements.... because the OFX data model is arranged such that you won't encounter a definition in a narrower scope overriding that of a wider scope; it's not really a method resolution order.

> Get to the god damned point already csingley

<ahem> So your `ticker` is in fact an attribute of the "superclass" `SECINFO` not the "subclass" `MFINFO`... and as we were exploring before you so rudely interrupted me, the `SECINFO` is in fact contained within the `MFINFO` as an attribute... so your attempts to set the `ticker` attribute on the `MFINFO` are of course doomed to fail; you're just creating a new plain-jane instance attribute instead of manipulating the data descriptors I created to embody the structure of the OFX data model.... and since you're just fabricating this purported `MFINFO.ticker` attribute out of whole cloth, when it doesn't exist within the OFX spec, why of course it's not going to be rendered by the reverse transform.

Clear as mud?  Dude just read the freaking object model; I took some pains to make it as transparent as humanly possible, trivially able to be validated against the actual spec.

[`SECINFO`](https://github.com/csingley/ofxtools/blob/master/ofxtools/models/invest/securities.py#L86-L99) vs. [`MFINFO`](https://github.com/csingley/ofxtools/blob/master/ofxtools/models/invest/securities.py#L148-L171)

Still reading?  Where are you going?  There are such wonders contained within this deprecated, nigh-obsolete protocol.  Come back, I have so many things to tell you!

**Comment by pallavg** [comment]
Hi @csingley 
1. Thanks for the help/explanation. It is clear. 
2. I did switch to Xpath support to make the change and it did work. I was able to feed it to beancount-import to get the transactions imported.
3. Thanks for this great library.
